subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_psa.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <errno.h>
+#include <zephyr/init.h>
+#include <psa/crypto.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(fp_crypto, CONFIG_FP_CRYPTO_LOG_LEVEL);
+
+#include "fp_crypto.h"
+
+
+int fp_crypto_sha256(uint8_t *out, const uint8_t *in, size_t data_len)
+{
+	size_t hash_len = 0;
+	psa_status_t status = psa_hash_compute(PSA_ALG_SHA_256, in, data_len,
+					       out, FP_CRYPTO_SHA256_HASH_LEN, &hash_len);
+
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_hash_compute failed (err: %d)", status);
+		return -EIO;
+	}
+
+	if (hash_len != FP_CRYPTO_SHA256_HASH_LEN) {
+		LOG_ERR("Invalid psa_hash_compute output len: %zu", hash_len);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static psa_key_id_t import_hmac_sha256_key(const uint8_t *data, size_t len)
+{
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+
+	psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_SIGN_HASH);
+	psa_set_key_lifetime(&key_attr, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attr, PSA_ALG_HMAC(PSA_ALG_SHA_256));
+	psa_set_key_type(&key_attr, PSA_KEY_TYPE_HMAC);
+	psa_set_key_bits(&key_attr, len * CHAR_BIT);
+
+	psa_status_t status = psa_import_key(&key_attr, data, len, &key_id);
+
+	psa_reset_key_attributes(&key_attr);
+
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_import_key failed (err: %d)", status);
+		key_id = PSA_KEY_ID_NULL;
+	}
+
+	return key_id;
+}
+
+int fp_crypto_hmac_sha256(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *hmac_key,
+			  size_t hmac_key_len)
+{
+	int err = 0;
+	size_t olen = 0;
+	psa_key_id_t hmac_key_id;
+	psa_status_t status;
+
+	hmac_key_id = import_hmac_sha256_key(hmac_key, hmac_key_len);
+	if (hmac_key_id == PSA_KEY_ID_NULL) {
+		LOG_ERR("import_hmac_sha256_key failed");
+		return -EIO;
+	}
+
+	status = psa_mac_compute(hmac_key_id, PSA_ALG_HMAC(PSA_ALG_SHA_256), in, data_len,
+				 out, FP_CRYPTO_SHA256_HASH_LEN, &olen);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_mac_compute failed (err: %d)", status);
+		err = -EIO;
+	}
+
+	if (!err && (olen != FP_CRYPTO_SHA256_HASH_LEN)) {
+		LOG_ERR("Invalid psa_mac_compute output length: %zu", olen);
+		err = -EIO;
+	}
+
+	status = psa_destroy_key(hmac_key_id);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_destroy_key failed (err: %d)", status);
+	}
```

Comment:
Initially I wanted to avoid returning error here, because crypto operation itself was successful (only the cleanup fails).

Reply:
The test seems still marked as successful even if an error log is emitted, so avoiding passing an error code here could make us overlook `psa_destroy_key ` failures during unit test execution. Since we actually don't know what could be the root cause of the `psa_destroy_key` failure (guess it may even depend on PSA crypto implementation), I would prefer not to trigger assertion failure or `k_panic` here. Adding a test-specific error forwarding may not be a proper approach too (it modifies code execution only for testing). I think I would simply return another error code here to ensure application/test is aware of the problem (for application it should not happen anyway and application may not be able to do anything about it).

Reply:
Discussed offline, we will use `-ECANCELED` in that case.